### PR TITLE
improve ts->mp4 convertor build script

### PR DESCRIPTION
**Pull request by lilachmaliniak**
merge plat-7391. Update and extend build scripts.

### CODE REVIEW

File: liveRecorder/scripts/ts2mp4_build.sh [hunk 1]
```diff
@@ -1,58 +1,69 @@
...
+	BUILD_PATH=../${TARGET}
+	BUILD_FFMPEG_PATH=../../node_addons/FormatConverter/build/FFmpeg
+fi
 
-[ -d "$currentDir/../bin/${OS}" ] || mkdir -p "$currentDir/../bin/${OS}"
+if [ -e $CONVERTOR_DIR ]; then
+	echo "CONVERTOR_DIR=${CONVERTOR_DIR}, running make to create ${TARGET}"
+	echo uname=$(uname)
+	uname=$(uname)
+	OS=darwin
```
Comment (kalturaguy) [inline liveRecorder/scripts/ts2mp4_build.sh]: please remove OS dependency... since we are compling anyway
also remove bin files from repo

File: liveRecorder/scripts/ts2mp4_build.sh [hunk 2]
```diff
@@ -1,58 +1,69 @@
...
-   ;;
-esac
\ No newline at end of file
+	pushd $BUILD_PATH
+		mkdir -p obj
+		[ -e  $BUILD_PATH/${TARGET} ] && rm -f ${TARGET}
+		[ -e  obj/${TARGET} ] && rm -f obj/${TRAGET}
+		BASH_VERSION="${BASH_VERSINFO[0]}.${BASH_VERSINFO[1]}"
+		echo BASH_VERSION=${BASH_VERSION}
+		if ((BASH_VERSINFO[0] >= 4)); then
```
Comment (kalturaguy) [inline liveRecorder/scripts/ts2mp4_build.sh]: this won't work on all OS so I rather not even check this

File: liveRecorder/scripts/ts2mp4_build.sh [hunk 3]
```diff
@@ -1,58 +1,65 @@
...
+BUILD_FFMPEG_PATH=${WEBAPPDIR}/ffmpeg/ffmpeg-3.0
+ROOT_PATH=.
+TARGET=ts_to_mp4_convertor
+BIN_DIR=${ROOT_PATH}/../bin/
+BUILD_FFMPEG_PATH=
 
+if [ -e $PREPDIR ]; then
+	ROOT_PATH==${PREPDIR}/${APP_VERSION}
+	CONVERTOR_DIR=${ROOT_PATH}/${TARGET}
+	BUILD_PATH=${CONVERTOR_DIR}/obj
```
Comment (kalturaguy) [inline liveRecorder/scripts/ts2mp4_build.sh]: BUILD_PATH=${CONVERTOR_DIR}/obj
is the same for both cases

File: liveRecorder/Tasks/ConcatinationTask.py [hunk 4]
```diff
@@ -28,7 +28,7 @@ class ConcatenationTask(TaskBase):
     token_url_template = nginx_host + ":" + nginx_port +"/dc-0/recording/hls/p/0/e/{0}/"
     os_name = platform.system().lower()
```
Comment (kalturaguy) [inline liveRecorder/Tasks/ConcatinationTask.py]: we can remove os_name and import platform

File: node_addons/FormatConverter/setenv.sh [hunk 5]
```diff
@@ -1,118 +1,149 @@
...
+	FFMPEG_DEV_PATH=$PREPDIR/node_addons/FormatConverter/build/FFmpeg
+else
+	FFMPEG_DEV_PATH=$ROOT_PATH/build/FFmpeg
+fi
 
+[ -d "$FFMPEG_BUILD_PATH" ] || mkdir -p "FFMPEG_BUILD_PATH"
 
-os_name=`uname`
 
 function makeFFmpeg()
```
Comment (kalturaguy) [inline node_addons/FormatConverter/setenv.sh]: this should be extracted to another script .

File: node_addons/FormatConverter/setenv.sh [hunk 6]
```diff
@@ -1,118 +1,149 @@
...
+			        shift
+        ;;
+        -m | --mode )
+                    shift
+                    [ $1 = 'debug' ] && BUILD_CONF=Debug || BUILD_CONF=Release
+			        echo "-m | --mode $1"
+			        shift
+		;;
+        -h | --help )
+			        echo "-p | --path [full path], FormatConverter bin path"
```
Comment (kalturaguy) [inline node_addons/FormatConverter/setenv.sh]: why give a different format convertor path?

the idea is to give FFMPEG source location

File: node_addons/FormatConverter/setenv.sh [hunk 7]
```diff
@@ -1,118 +1,149 @@
...
-    configFileName=$ffmpegDir/lastConfigure
+	    debug_specifics=
+	    [ "$BUILD_CONF" = "Debug" ] &&  debug_specifics='--enable-debug --disable-optimizations'
 
+	    configFileName=./lastConfigure
 
-    confCmd="./configure --disable-everything --disable-doc --enable-protocol=file --enable-demuxer=mpegts --enable-muxer=rtp_mpegts --enable-parser=h264 --enable-parser=aac --enable-muxer=mp4 --enable-zlib --enable-bsf=aac_adtstoasc --enable-decoder=aac --enable-decoder=h264 --enable-muxer=flv --enable-protocol=rtmp --enable-encoder=libmp3lame $debug_specifics"
+	    confCmd="./configure --disable-everything --disable-doc --enable-protocol=file --enable-demuxer=mpegts --enable-muxer=rtp_mpegts --enable-parser=h264 --enable-parser=aac --enable-muxer=mp4 --enable-zlib --enable-bsf=aac_adtstoasc --enable-decoder=aac --enable-decoder=h264 --enable-muxer=flv --enable-protocol=rtmp --enable-encoder=libmp3lame $debug_specifics"
 
+	    [ "$os_name" == "Linux" ] && confCmd="$confCmd --enable-pic"
```
Comment (kalturaguy) [inline node_addons/FormatConverter/setenv.sh]: i think some os can return linux and Linux, please check

File: node_addons/FormatConverter/setenv.sh [hunk 8]
```diff
@@ -1,118 +1,149 @@
...
-    ;;
-'Darwin')
-    cp `pwd`/build/$build_conf/FormatConverter.dylib "`pwd`/../../bin/darwin/FormatConverter.node$debugExt"
-    ;;
-*) ;;
-esac
\ No newline at end of file
+echo "cp ${FFMPEG_BUILD_PATH}/${FORMAT_CONVERTER_BIN} ${BIN_PATH}/FormatConverter.node$debugExt"
+
+cp "${FFMPEG_BUILD_PATH}/${FORMAT_CONVERTER_BIN}" "${BIN_PATH}/FormatConverter.node$debugExt"
```
Comment (kalturaguy) [inline node_addons/FormatConverter/setenv.sh]: 1.please delete the FormatConverter.node from git
2.you didn't fix the actual javascript code that loads the FormatConverter.node with the 'new' path

File: liveRecorder/Tasks/ConcatinationTask.py (hunk 4 above)
Comment (lilachmaliniak) [inline liveRecorder/Tasks/ConcatinationTask.py]: done

File: lib/MP4WriteStream.js [hunk 9]
```diff
@@ -17,7 +17,7 @@ var preserveOriginalHLS = config.get('preserveOriginalHLS').enable;
 var Q = require('q');
 var ErrorUtils = require('./utils/error-utils');
 
-function getArchAndPlatform()
+/*function getArchAndPlatform()
```
Comment (kalturaguy) [inline lib/MP4WriteStream.js]: delete

File: scripts/live_prep.sh [hunk 10]
```diff
@@ -0,0 +1,181 @@
+# !/bin/bash
```
Comment (kalturaguy) [inline scripts/live_prep.sh]: please rename to build_addon.sh

File: scripts/live_prep.sh [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+if [ "$BUILD_CONF" = "Debug" ]; then
+	ADDONS_BUILD_PATH=build/Debug
+    echo "target config: debug"
+else
+    echo "target config: release"
+fi
+
+echo "current path `pwd`"
+
+if [ "$BUILD_TARGET" = "ffmpeg" ] || [ "$BUILD_TARGET" = "all" ]; then
```
Comment (kalturaguy) [inline scripts/live_prep.sh]: build_ffmpeg should not be here, we will call it outside the context of this script

File: scripts/live_prep.sh [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+	echo "sh $SCRIPT_PATH/build_ffmpeg.sh"
+
+	sh $SCRIPT_PATH/build_ffmpeg.sh
+
+	echo "*******************************************************************************"
+	echo "* finished building ffmpeg                                                    *"
+	echo "*******************************************************************************"
+fi
+
+if [ "$BUILD_TARGET" = "node_addons" ] || [ "$BUILD_TARGET" = "all" ]; then
```
Comment (kalturaguy) [inline scripts/live_prep.sh]: plesae remvoe the BUILD_TARGET

File: scripts/live_prep.sh [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+OS=`uname`
+echo "OS=$OS"
+
+contains_element () {
+  local e
+  for e in "${@:2}"; do [[ "$e" == "$1" ]] && return 0; done
+  return 1
+}
+
+while [ "$1" != "" ]; do
```
Comment (kalturaguy) [inline scripts/live_prep.sh]: this entire while code is a bit out of scope....   we can just check number of params and return error if something is missing

File: scripts/build_addons.sh [hunk 14]
```diff
@@ -0,0 +1,55 @@
+# !/bin/bash
+
+if [ "$#" -eq 0 ]; then
+	echo "usage build_addons <product path> <release/debug - optional>"
+	exit
+fi
+
+PRODUCT_ROOT_PATH=$1
```
Comment (kalturaguy) [inline scripts/build_addons.sh]: PRODUCT_ROOT_PATH  should not be something you pass, it's a relative directory

File: scripts/build_addons.sh [hunk 15]
```diff
@@ -0,0 +1,55 @@
+# !/bin/bash
+
+if [ "$#" -eq 0 ]; then
+	echo "usage build_addons <product path> <release/debug - optional>"
```
Comment (kalturaguy) [inline scripts/build_addons.sh]: ffmpeg base path missing?
build path missing (optional)

product path should never be passed

File: scripts/build_addons.sh [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+	exit
+fi
+
+PRODUCT_ROOT_PATH=$1
+BUILD_CONF=$2
+ADDONS_PATH=$PRODUCT_ROOT_PATH/node_addons/FormatConverter
+FORMAT_CONVERTER_BIN=FormatConverter.so
+
+if [ "$BUILD_CONF" = "debug" ]; then
+	BUILD_CONF=DEBUG
```
Comment (kalturaguy) [inline scripts/build_addons.sh]: is it caps lock in purpose?

File: lib/MP4WriteStream.js (hunk 9 above)
Comment (lilachmaliniak) [inline lib/MP4WriteStream.js]: deleted

File: scripts/live_prep.sh (hunk 13 above)
Comment (lilachmaliniak) [inline scripts/live_prep.sh]: chaged

File: scripts/live_prep.sh (hunk 12 above)
Comment (lilachmaliniak) [inline scripts/live_prep.sh]: done

File: liveRecorder/scripts/build_ts2mp4_convertor.sh [hunk 17]
```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
+
+if [ "$#" -lt 2 ]; then
+	echo "usage build_ts2mp4_convertor <ffmpeg build path> <product path>"
+	exit
```
Comment (jessp01) [inline liveRecorder/scripts/build_ts2mp4_convertor.sh]: Should always exit with a non zero RC in the event of an error.

File: liveRecorder/scripts/build_ts2mp4_convertor.sh [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+PRODUCT_ROOT_PATH=$2
+TARGET=ts_to_mp4_convertor
+CONVERTOR_DIR=${PRODUCT_ROOT_PATH}/${TARGET}
+
+[ `uname` = "Linux" ] && OS="linux" || OS="darwin"
+
+export FFMPEG_BUILD_PATH
+export OS
+
+echo "OS=$OS"
```
Comment (jessp01) [inline liveRecorder/scripts/build_ts2mp4_convertor.sh]: What is OS being used for?

File: liveRecorder/scripts/build_ts2mp4_convertor.sh [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+
+if [ -e $CONVERTOR_DIR ]; then
```
Comment (jessp01) [inline liveRecorder/scripts/build_ts2mp4_convertor.sh]: If you want to check if a dir is readable, always use -r, for writeable, use -w.
-e is not enough and is almost never useful. For example:
```
jess@jessex:~$ whoami 
jess
jess@jessex:~$ ll -d /root/
drwx------ 67 root root 28672 Jun 22 13:08 /root//
jess@jessex:~$ if [ -e /root ];then echo exists;fi
exists
jess@jessex:~$ if [ -r /root ];then echo readable;fi
jess@jessex:~$ if [ -w /root ];then echo writeable;fi
```
So yes, /root exists but the jess user can't even read it so what good does it do that it exists?

File: liveRecorder/scripts/build_ts2mp4_convertor.sh [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+	pushd $CONVERTOR_DIR
+		mkdir -p obj
+		[ -e  ${TARGET} ] && rm -f ${TARGET}
```
Comment (jessp01) [inline liveRecorder/scripts/build_ts2mp4_convertor.sh]: Since you're calling rm with -f, no point in checking whether $TARGET exists. with -f, rm will return RC 0 even if the file does not exist.

File: liveRecorder/scripts/build_ts2mp4_convertor.sh [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+		[ -e  obj/${TARGET} ] && rm -f obj/${TARGET}
+
+		echo "starting to build ${TARGET}"
+
+		make
+
+		if [ -s  obj/${TARGET} ] ; then
```
Comment (jessp01) [inline liveRecorder/scripts/build_ts2mp4_convertor.sh]: Instead of checking whether obj/${TARGET}'s size is bigger than zero, check the RC from make with:
```
if [ $? -eq 0 ]
```
Obviously, if make exited with a non zero RC, we failed.

File: liveRecorder/scripts/build_ts2mp4_convertor.sh [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+			echo "**************************************************************************************"
+			echo "${TARGET} was built successfully, copying to bin folder"
+			echo "**************************************************************************************"
+			mkdir -p bin
+			echo "cp obj/${TARGET} bin"
+			cp obj/${TARGET}  ../bin
```
Comment (jessp01) [inline liveRecorder/scripts/build_ts2mp4_convertor.sh]: You shouldn't be manually calling cp from the shell script, the Makefile should have a "install" target.

File: scripts/build_addon.sh [hunk 23]
```diff
@@ -0,0 +1,64 @@
+# !/bin/bash
+
+if [ "$#" -eq 0 ]; then
+	echo "usage build_addon <product path> <release/debug - optional>"
+	exit
```
Comment (jessp01) [inline scripts/build_addon.sh]:
```
if [ "$#" -lt 1 ]; then
```
makes more sense. Also, as a convention, optional args are marked with "[]", like so:
```
echo "usage build_addon <product path> [release/debug]"
```
Instead of 'build_addon', using $1.
Lastly, should always exit with a non zero RC in the event of an error.

File: scripts/build_addon.sh [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+PRODUCT_ROOT_PATH=$1
+BUILD_CONF=
+ADDON_PATH=$PRODUCT_ROOT_PATH/node_addons/FormatConverter
+FORMAT_CONVERTER_BIN=FormatConverter.so
+OS=`uname`
+echo "OS=$OS"
+
+[ "$#" -eq 2 ] && BUILD_CONF=$2 || BUILD_CONF=Release
+
+if [ "$BUILD_CONF" = "debug" ]; then
```
Comment (jessp01) [inline scripts/build_addon.sh]: Why not just say you expect 'Release' || 'Debug' in the first place? rather than expect and check for lowercase only to ucf() it?

File: scripts/build_addon.sh [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+	echo "Release mode"
+fi
+
+echo "ADDON_PATH=$ADDON_PATH"
+
+pushd $ADDON_PATH
+
+	`which node-gyp` || npm install node-gyp -g
+
+	[ -d '/usr/local/lib/node_modules/nan' ] || npm install nan -g
```
Comment (jessp01) [inline scripts/build_addon.sh]: Never presume the global NPM modules dir is /usr/local/lib/node_modules.
Use:
```
npm list --depth 1 -g nan
```
and check the RC.

File: scripts/build_addon.sh [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+
+	gyp_args=''
+
+	echo "Installing NAN"
+	npm install -unsafe-perm nan
```
Comment (jessp01) [inline scripts/build_addon.sh]: Why are we install nan twice? one globally [line 31] and once locally?

File: scripts/build_ffmpeg.sh [hunk 27]
```diff
@@ -0,0 +1,86 @@
+# !/bin/bash
+
+if [ "$#" -lt 2 ]; then
```
Comment (jessp01) [inline scripts/build_ffmpeg.sh]: See my note on scripts/build_addon.sh and liveRecorder/scripts/build_ts2mp4_convertor.sh

File: scripts/build_ffmpeg.sh [hunk 28]
```diff
@@ -0,0 +1,86 @@
...
+FFMPEG_BUILD_PATH=$1
+PRODUCT_ROOT_PATH=$2
+ADDON_BUILD_PATH=$PRODUCT_ROOT_PATH/node_addons/FormatConverter/build/
+FFMPEG_SYMLINK=$ADDON_BUILD_PATH/FFmpeg
+BUILD_CONF=
+OS=`uname`
+
+echo "PRODUCT_ROOT_PATH=$PRODUCT_ROOT_PATH"
+
+[ -d $FFMPEG_BUILD_PATH ] || mkdir -p "$FFMPEG_BUILD_PATH"
```
Comment (jessp01) [inline scripts/build_ffmpeg.sh]: mkdir -p already means it won't fail if it exists so -d is not needed.

File: scripts/build_ffmpeg.sh [hunk 29]
```diff
@@ -0,0 +1,86 @@
...
+
+[ "$#" -eq 3 ] && BUILD_CONF=$3 || BUILD_CONF=Release
+
+if [ "$BUILD_CONF" = "debug" ]; then
```
Comment (jessp01) [inline scripts/build_ffmpeg.sh]: See my note on scripts/build_addon.sh, same story...

File: scripts/build_ffmpeg.sh [hunk 30]
```diff
@@ -0,0 +1,86 @@
...
+echo "FFMPEG_BUILD_PATH=$FFMPEG_BUILD_PATH"
+echo "ADDON_BUILD_PATH=$ADDON_BUILD_PATH"
+
+echo "unlink $FFMPEG_SYMLINK"
+unlink $FFMPEG_SYMLINK
+
+
+if [ ! -d $FFMPEG_SYMLINK ]; then
+
+    curl -OL https://github.com/FFmpeg/FFmpeg/releases/download/n3.0/ffmpeg-3.0.tar.gz
```
Comment (jessp01) [inline scripts/build_ffmpeg.sh]: Set FFMPEG_VERSION and use it rather than repeating the literal '3.0' six times. Much easier to maintain.

File: scripts/build_ffmpeg.sh [hunk 31]
```diff
@@ -0,0 +1,86 @@
...
+    mv ./ffmpeg-3.0.tar.gz /var/tmp/ffmpeg-3.0.tar.gz
```
Comment (jessp01) [inline scripts/build_ffmpeg.sh]: if you want the file saved to /var/tmp, why not:
```
curl -L https://github.com/FFmpeg/FFmpeg/releases/download/n3.0/ffmpeg-3.0.tar.gz -o /tmp/ffmpeg-3.0.tar.gz
```
instead of first downloading to CWD and then moving it?

File: scripts/build_ffmpeg.sh [hunk 32]
```diff
@@ -0,0 +1,86 @@
...
+    echo "current path `pwd`"
+
+    debug_specifics=
+    [ "$BUILD_CONF" = "Debug" ] &&  debug_specifics='--enable-debug --disable-optimizations'
+
+    configFileName=./lastConfigure
+
+    confCmd="./configure --disable-everything --disable-doc --enable-protocol=file --enable-demuxer=mpegts --enable-muxer=rtp_mpegts --enable-parser=h264 --enable-parser=aac --enable-muxer=mp4 --enable-zlib --enable-bsf=aac_adtstoasc --enable-decoder=aac --enable-decoder=h264 --enable-muxer=flv --enable-protocol=rtmp --enable-encoder=libmp3lame $debug_specifics"
+
+    [ $OS == "Linux" ] && confCmd="$confCmd --enable-pic"
```
Comment (jessp01) [inline scripts/build_ffmpeg.sh]: Why is this needed? ffmpeg correctly uses GNU's autotools to determine which build flags [CFLAGS and LDFLAGS]  are needed per OS and arch.

File: liveRecorder/scripts/build_ts2mp4_convertor.sh (hunk 18 above)
Comment (lilachmaliniak) [inline liveRecorder/scripts/build_ts2mp4_convertor.sh]: It is for the makfie

Comment (jessp01) [inline liveRecorder/scripts/build_ts2mp4_convertor.sh]: So put it directly there instead.

File: liveRecorder/scripts/build_ts2mp4_convertor.sh (hunk 19 above)
Comment (lilachmaliniak) [inline liveRecorder/scripts/build_ts2mp4_convertor.sh]: ok.tx

File: liveRecorder/scripts/build_ts2mp4_convertor.sh (hunk 21 above)
Comment (lilachmaliniak) [inline liveRecorder/scripts/build_ts2mp4_convertor.sh]: done

File: liveRecorder/scripts/build_ts2mp4_convertor.sh (hunk 20 above)
Comment (lilachmaliniak) [inline liveRecorder/scripts/build_ts2mp4_convertor.sh]: updated

File: liveRecorder/scripts/build_ts2mp4_convertor.sh (hunk 22 above)
Comment (lilachmaliniak) [inline liveRecorder/scripts/build_ts2mp4_convertor.sh]: done

File: scripts/build_addon.sh (hunk 23 above)
Comment (lilachmaliniak) [inline scripts/build_addon.sh]: done

File: packager/bin/build_nginx.sh [hunk 33]
```diff
@@ -1,5 +1,11 @@
 # !/bin/bash
 
+if [ "$#" -lt 1 ]; then
+	echo "usage nginx <version>"
+	exit
```
Comment (jessp01) [inline packager/bin/build_nginx.sh]: See my notes on all the other scripts. You need to exit with a non zero RC.

Comment (lilachmaliniak) [inline packager/bin/build_nginx.sh]: done

Comment (jessp01) [inline packager/bin/build_nginx.sh]: I still see exit without an RC:
https://github.com/kaltura/liveDVR/blob/e4c02250d02be026dc7d10806849e1c97d8c91c6/packager/bin/build_nginx.sh#L5

File: deployment/get_bins.sh [hunk 34]
```diff
@@ -0,0 +1,56 @@
...
+#  ORGANIZATION: Kaltura, inc.
+#       CREATED: June 26, 2017
+#      REVISION:  ---
+#===============================================================================
+
+if [ "$#" -lt 1 ]; then
+	echo "usage get_bins <version> [live root path]"
+	exit 1
+fi
+
```
Comment (kalturaguy) [inline deployment/get_bins.sh]: i think this could be written better with for loop and a list of source and targets...

**Comment by lilachmaliniak** [comment]
update scripts and files to support on-prem installation (plat-7391)
